pkg/api/types.go

```diff
@@ -1205,6 +1221,13 @@ type KeyToPath struct {
 	Mode *int32
 }
 
+// Local represents a directory with specific topology contraints
+type LocalVolumeSource struct {
+	// The full path to the volume on the node
+	// For alpha, this path must be a file-based volume
+	Path string
```

Comment:
Let's be very crisp about the purpose of the information we are trying to find a home for in the API. Doing so will help make it easier to justify where in the API it makes sense.

The way I see it, we need to differentiate between two pieces of information:
1. Generic volume information
    * Information Kubernetes must understand and operate on in order to complete the volume setup process.
2. Plugin specific information
    * Relevant to individual volume plugins and (likely) different between volume plugins.
    * Opaque to Kubernetes: only plugin needs to understand this.
    * In the future with CSI plugins this will likely no longer be part of the Kubernetes API, and will be demoted to opaque pass through key/value pairs.

In this thread we have the following pieces of information:
1) Volume consumption information: block OR mount
    * Indicates if volume will be consumed as a block device OR as mounted volume.
    * If block device, no further information required about consumption.
    * If mounted volume, must also specify:
      * File system
      * Additional mount flags
    * *This is "Generic volume information" as defined above, because k8s must be aware of this to figure out how to properly surface the volume in the container (block device or mounted directory).*
2) Volume connection information: information required by plugin to identify volume
    * This is "plugin specific information" as defined above, because k8s does not care what it is (only the volume plugin does).
    * For "local storage" volume plugin, this includes  "volume source" information:
      * For block device local storage, this is the "block device path".
      * For mounted volumes local storage, this is the "path to mounted directory".

Based on this breakdown, here are my thoughts on @msau42's options:
> 1. Having a file struct and a block struct within the same volume source object. None of the other volume objects do this, so it looks inconsistent.

Not a good option. It requires the k8s to understand and use volume plugin specific fields to figure out how to surface the volume (these fields should only be used by the volume plugin)

> 2. Overload Path field and use PV.VolumeType field to distinguish between the two. The Path field having different behaviors/semantics depending on the type may be difficult for users to understand.

Ok option. Since the "source/connection information" is inside the plugin specific fields, and the "consumption information (block or mount)" is in the generic k8s volume fields. 

> 3. Have two completely separate volume plugins, one for local file volumes, and one for local block volumes. This is more consistent with the current semantics (kinda like rbd and cephfs). We are trying not to hardcode the Local volume type into the scheduler and pv controller, so adding a block volume plugin in the future should work as long as it uses the same general topology mechanisms.

Ok as long as we leave the mount vs block info in PV/PVC (since it is required by Kubernetes), but doing so AND having separate "block only" and "mount only" local storage plugins is very ugly.

So I perfer @msau42's second option, but I understand @jsafrane's concerns, and would also be ok with his earlier proposal to break "path" into two volume specific fields inside the local storage volume: `LocalVolumeSource.PathToDevice` and `LocalVolumeSource.PathToMountedDirectory` (although that looks a little ugly too).